test, ci: add --v1transport option, add --v2transport  to a CI task

mzumsande opened the pull request:
This suggests a strategy to run the functional tests with both v1 and v2 transport in the CI.

**Status Quo:**
There is both the global `--v2transport` option for the `test_runner.py` (not enabled by default), plus the possibility to specify `--v2transport` for particular tests, which is used for a handful of tests. Currently, when running `test_runner.py --v2transport`, these tests are run twice with the same `--v2transport` configuration, as has been noted in https://github.com/bitcoin/bitcoin/pull/29358#discussion_r1485626063, which is wasteful.

**Suggested Change:**
Fix this by adding a `--v1transport` option and using it in `test_runner.py`, so that irrespective of the global `--v2transport` flag, the  tests that run twice use v1 in one run and v2 in the other.
Also add `--v2transport` to one CI task (`multiprocess, i686, DEBUG`).
This means, that for each CI task, the majority of functional tests will run once using the global `--v2transport` option if provided, while a few selected tests will always run two times, once with `v1` and once with `v2`.

**Rationale:**
A simpler alternative would have been to remove all test-specific `--v2transport` commands from `test_runner.py` and just enable `--v2transport` option for a few CI tasks. I didn't do that because it would have meant that v2 would never be running in the CI for some platforms, and also be run a lot less locally by users and devs (who would have to actively enable the `--v2transport` option).

mzumsande committed:
test: add -v1transport option and use it in test_runner

This option beats the --v2transport option and is meant to be used in
test_runner.py.
It applies these to a few tests that are particulary interesting
in terms of the transport type.
This ensures that these tests arei always run with both v1 and v2, irrespective of
whether the global --v2transport test_runner option is set or not.

mzumsande committed:
ci: add --v2transport to an existing CI job

kevkevinpal commented:
Concept ACK 

it seems like we followed a similar approach for `--legacy-wallet` and `--descriptors`

fanquake commented:
Could rebase / update PR description. I think this could also be worthwhile having in 27.

mzumsande commented:
updated PR description, it's rebased already.

Since this runs a large numbers of tests in the CI with v2 that previously weren't, there is a certain risk of new intermittent failures (though I didn't counter any locally and the CI runs so far didn't either).
If it's meant for 27, I'll add a temporary no-merge commit to enable v2 in each CI task running the functional tests - just to have a few more runs with different CIs, I'll remove that commit again this evening.

mzumsande committed:
test: ignore --v2transport for older versions instead of asserting

Otherwise, a run with
test_runner.py --v2transport=1 --previous-releases --extended
would hit the removed assert for wallet_backwards_compatibility.py

mzumsande commented:
>Since this runs a large numbers of tests in the CI with v2 that previously weren't, there is a certain risk of new intermittent failures (though I didn't encounter any locally and the CI runs so far didn't either).
If it's meant for 27, I'll add a temporary no-merge commit to enable v2 in each CI task running the functional tests - just to have a few more runs with different CIs, I'll remove that commit again this evening.

ok, no intermittent bug was found that way, but a non-intermittent problem, fixed in 3a25a575f0c455b50abf95d85c0ba8de3cf53a87

stratospher reviewed (commented):
ACK ecc036c.

tdb3 commented:
ACK for ecc036c5d63fb4bdff5caeede88baeb85bf62b05.

### Checks
Ran a quick set of sanity checks.  Received expected results in every check.

Added log.info() statements to BitcoinTestFramework.setup() showing v1transport and v2transport status, as well as in TestNode's constructor showing v2transport status.  Checked test_framework logs for state of v1transport and v2transport.

__Not globally specifying `--v2transport`:__

`test/functional/test_runner.py --cachedir=/mnt/tmp/cache --tmpdir=/mnt/tmp -u --nocleanup test/functional/p2p_block_sync*`
`p2p_block_sync.py --v1transport` passed and logs showed v1transport enabled and v2transport disabled for the test (including for each node involved).
`p2p_block_sync.py --v2transport` passed and logs showed v1transport disabled and v2transport enabled for the test (including for each node involved).

__Globally specifying `--v2transport`:__

`test/functional/test_runner.py --cachedir=/mnt/tmp/cache --tmpdir=/mnt/tmp -u --v2transport --nocleanup test/functional/p2p_block_sync*`
`p2p_block_sync.py --v1transport` passed and logs showed v1transport enabled and v2transport disabled for the test (including for each node involved).
`p2p_block_sync.py --v2transport` passed and logs showed v1transport disabled and v2transport enabled for the test (including for each node involved).
 

`v2transport` is now enabled by default (PR #29347).  Initially, I was considering suggesting v2transport being enabled by default for tests, however the suggestion (in https://github.com/bitcoin/bitcoin/pull/29358#discussion_r1486473051) to default to v1 now and switch to v2 after more v2 adoption (e.g. when the majority of the network is using v2) makes more sense.  For now, it seems that using v2 would require either the global setting (`test_runner.py --v2transport`) or explicit declaration (`--v2transport`) in the test in `BASE_SCRIPTS` (so new tests should be explicit).  A future PR could introduce defaulting to v2 for tests.

vasild commented on test/functional/test_framework/test_framework.py:
```diff
@@ -191,6 +191,8 @@ def parse_args(self):
         parser.add_argument("--timeout-factor", dest="timeout_factor", type=float, help="adjust test timeouts by a factor. Setting it to 0 disables all timeouts")
         parser.add_argument("--v2transport", dest="v2transport", default=False, action="store_true",
                             help="use BIP324 v2 connections between all nodes by default")
+        parser.add_argument("--v1transport", dest="v1transport", default=False, action="store_true",
```
nit, in the commit message of 547aacff08a2a5d786cb0fa6c7bae022ea651f8c `test: add -v1transport option and use it in test_runner`:
`s/-v1transport/--v1transport/`
`s/arei/are/`

vasild commented on test/functional/test_framework/test_framework.py:
```diff
@@ -206,6 +208,8 @@ def parse_args(self):
         config = configparser.ConfigParser()
         config.read_file(open(self.options.configfile))
         self.config = config
+        if self.options.v1transport:
+            self.options.v2transport=False
```
v1 overriding v2 is somewhat non intuitive (v2 overriding v1 is as well). What we really want here is a local option overriding a global one, regardless of which one is v1 and v2. Can we have this somehow? Maybe have `test_runner.py` when ran like e.g. `test_runner.py --v1` and when picking entries from `BASE_SCRIPTS[]` to check that an entry like `p2p_timeouts.py --v2` already contains a conflicting option and omit that entry altogether?

Also, it would be best if an individual test is ran like `test/functional/p2p_fingerprint.py --v1 --v2` to emit an error and refuse to start (e.g. "conflicting options given") instead of running in `v1` mode.

vasild reviewed (approved):
ACK ecc036c5d63fb4bdff5caeede88baeb85bf62b05

Looks good as it is, but I wonder if we can do better, see [the comment below](https://github.com/bitcoin/bitcoin/pull/29483#discussion_r1517425505).

achow101 commented:
ACK ecc036c5d63fb4bdff5caeede88baeb85bf62b05

MarcusMWilliams commented:
Pickup # 1-225-436-2680

On Mon, Mar 11, 2024, 9:46 AM Martin Zumsande ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/functional/test_framework/test_framework.py
> <https://github.com/bitcoin/bitcoin/pull/29483#discussion_r1519837885>:
>
> > +        if self.options.v1transport:
> +            self.options.v2transport=False
>
> Concept ACK for a possible follow-up - though we should probably be
> careful with introspection of args so that we don't create a second monster
> like the bitcoind args parser for the functional tests.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/29483#discussion_r1519837885>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASHJW5ZMGJRWBLWJV2WGTEDYXXGVPAVCNFSM6AAAAABD22BS66VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMRYGAZDKMBWGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
